### PR TITLE
Implementation of advance() for parser in uBPF backend

### PR DESCRIPTION
**Pull request by tatry**
I've implemented `advance()` in the uBPF parser. This allows to skip some bits without need to define headers.

In this implementation parser advances the packet cursor. When the packet is too short, it goes to reject state.

I think that @osinstom should be assigned.

### CODE REVIEW

File: backends/ubpf/ubpfParser.cpp [hunk 1]
```diff
@@ -335,6 +337,31 @@ UBPFStateTranslationVisitor::compileLookahead(const IR::Expression* destination)
     builder->endOfStatement(true);
 }
 
+void
+UBPFStateTranslationVisitor::compileAdvance(const IR::Expression* expr) {
+    auto type = state->parser->typeMap->getType(expr);
+    auto etype = UBPFTypeFactory::instance->create(type);
+    const char * tmpVarName = "__tmp_9507";  // need to be some random name
```
Comment (mbudiu-vmw) [inline backends/ubpf/ubpfParser.cpp:344]: Use the refMap, it can generate fresh unique names.

File: backends/ubpf/ubpfParser.cpp [hunk 2]
```diff
@@ -367,6 +394,16 @@ bool UBPFStateTranslationVisitor::preorder(const IR::MethodCallExpression* expre
                 compileExtract(expression->arguments->at(0)->expression);
                 return false;
             }
+
+            if (extMethod->method->name.name == p4lib.packetIn.advance.name) {
+                if (expression->arguments->size() != 1) {
+                    ::error("Exactly one argument expected %1%", expression);
```
Comment (mbudiu-vmw) [inline backends/ubpf/ubpfParser.cpp:400]: Use the version of error with an additional argument of type ErrType. Or perhaps call modelError if there is such a function.

**Comment by tatry** [comment]
I think it would be better if @osinstom will review the code.

**Comment by osinstom** [comment]
LGTM, I think we can merge.

**Pull request merged**